File: Task 2/matrix_mul_int.c
/**
 * @file matrix_mul.c
 * 
 * @author Kowsik Nandagopan D (22250016)
 * @brief
 * A program to evaluvate exection time of simple multiplication operation.
 * @version 0.1
 * @date 2022-08-25
 * 
 * 
 */

#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <stdint.h>

double diff_timespec(const struct timespec *time_a, const struct timespec *time_b) {
    // Function to find difference between two timespec objects
    return (time_a->tv_sec - time_b->tv_sec) + (time_a->tv_nsec - time_b->tv_nsec) / 1000000000.0;
}

double matrix_mul(int N) {

    // Defining time variables
    struct timespec start_time, end_time;
    
    // Declare new variables
    int i, j, k;

    // Dynamically allocating space for the matrix
    int* mat[N];
    for (i = 0; i < N; i++)
        mat[i] = (int*)malloc(N * sizeof(int));

    // Result matrix initialization
    int* result[N];
    for (i = 0; i < N; i++)
        result[i] = (int*)malloc(N * sizeof(int));

    // Inititializing the matrix with numbers
    for (i=0; i<N; i++) {
        for (j=0; j<N; j++){
            mat[i][j] = 100;
        }
    }

    // Start time
    timespec_get(&start_time, TIME_UTC);
    // Matrix multiplication
    for (i=0; i<N; i++) {
        for (j=0; j<N; j++) {
            for (k=0; k<N; k++) {
                result[i][j] += mat[i][k] * mat[k][j];
            }
        }
    }
    // End Time
    timespec_get(&end_time, TIME_UTC);

    // Print the difference in time (Run time of matrix multiplication)
    double time_taken = diff_timespec(&end_time, &start_time);
    printf("Run time of matrix multiplication = %f seconds\n", time_taken);

    return time_taken;
}

void main() {
    
    
    // Declare variables
    int N;
    double time_taken_multiplication, time_taken_whole;

    // Defining time variables
    struct timespec start_time, end_time;
    
    // For N = 256, 512, 1024, 2048, 4096
    N=256;

    // Start time
    timespec_get(&start_time, TIME_UTC);

    // Function call
    time_taken_multiplication = matrix_mul(N);
    
    // End Time
    timespec_get(&end_time, TIME_UTC);
    time_taken_whole = diff_timespec(&end_time, &start_time);
    printf("Run time of program (N=%d) = %f seconds\n", N, time_taken_whole);
    printf("Ratio of time taken by matrix multiplication to the total program = %f\n", (time_taken_multiplication / time_taken_whole));
}

File: Task 2/MatrixMulInt.java

import java.text.DecimalFormat;

public class MatrixMulInt {

    private static double MatrixMul (int N) {
        // Function to multipy 2 matrix and compute run time

        // Initializing variables
        double time_taken = 0;
        long start_time = 0, end_time = 0;
        int i, j, k;

        // Initialize N * N Matrix
        int[][] mat = new int[N][N];
        int[][] result = new int[N][N];

        // Initial value inside each matrix
        for (i=0; i<N; i++) {
            for (j=0; j<N; j++) {
                mat[i][j] = 100;
                result[i][j] = 0;
            }
        }

        // start time
        start_time = System.nanoTime();

        // Matrix multiplication
        for (i=0; i<N; i++) {
            for (j=0; j<N; j++) {
                for (k=0; k<N; k++) {
                    result[i][j] += mat[i][k] * mat[k][j];
                }
            }
        }

        // End time
        end_time = System.nanoTime();

        // Time taken for matrix multiplication
        time_taken = (end_time - start_time) * 1E-9;
        return time_taken;
    }

    public static void main (String[] argv) {

        // For N = 256, 512, 1024, 2048, 4096
        int N = 4096;

        // Initializing some variables
        double time_matrix_multiplication = 0, time_taken_program = 0;
        DecimalFormat ft = new DecimalFormat("##.######");
        long start_time = 0, end_time = 0;
       
        // Start time
        start_time = System.nanoTime();

        // Function call
        time_matrix_multiplication =  MatrixMul(N);

        // End time
        end_time = System.nanoTime();

        // Time taken for the program
        time_taken_program = (end_time - start_time) * 1E-9;
        
        System.out.println("Run time of matrix multiplication = " + ft.format(time_matrix_multiplication) + " seconds");
        System.out.println("Run time of program (N="+ N +") = " + ft.format(time_taken_program) + " seconds");
        System.out.println("Ratio of time taken by matrix multiplication to the total program = " + ft.format(time_matrix_multiplication/time_taken_program));
    }
}
